src/recommendations/src/recommendations-service/experimentation/resolvers.py

```diff
@@ -351,6 +372,82 @@ def get_items(self, **kwargs):
 
         return echo_items
 
+
+class PersonalizeContextComparePickResolver(Resolver):
+    """ Provides personalized ranking of products from an Amazon Personalize campaign
+
+    The campaign must be trained using the Personalized-Ranking recipe
+    """
+    __personalize_runtime = boto3.client('personalize-runtime')
+
+    def __init__(self, **params):
+        with_context = params.get('with_context')
+        without_context = params.get('with_context')
+        self.with_resolver = PersonalizeRankingResolver(**params, context=with_context)
+        self.without_resolver = PersonalizeRankingResolver(**params, context=without_context)
+
+    def get_items(self, **kwargs):
+        """ Returns reranking items from an Amazon Personalize campaign trained with Personalized-Ranking recipe
+
+        Arguments:
+            user_id - ID for the user for which to rerank items (required for Personalized-Ranking recipe)
+            product_list - list of product IDs to rerank for the user
+        """
+        top_n = kwargs.get('num_results')
+
+        if top_n is None:
+            raise Exception('num_results is required')
+
+        log.debug('PersonalizeContextComparePickResolver - comparing personalized rankings...')
+        with_ranked = self.with_resolver.get_items(**kwargs)
+        without_ranked = self.without_resolver.get_items(**kwargs)
+        without_id_to_item = {item['itemId']: item for item in without_ranked}
+        with_id_to_item = {item['itemId']: item for item in with_ranked}
+        score_increases_with_discount = {item_id: with_id_to_item[item_id]['score'] / (0.01 + without_id_to_item[item_id]['score'])
+                                         for item_id in with_id_to_item}
+        # Let us get the items sorted according to this score:
+        discount_improve_sorted_item_ids = sorted(score_increases_with_discount.keys(),
+                                                  key=lambda item_id: score_increases_with_discount[item_id])
+
+        discount_improve_sorted_items = [with_id_to_item[item_id] for item_id in discount_improve_sorted_item_ids]
+
+        return discount_improve_sorted_items[:top_n]
+
+
+class RandomPickResolver(Resolver):
+    """ Picks random N products.
+    """
+    __personalize_runtime = boto3.client('personalize-runtime')
```

Comment:
Does not appear to be used.

Reply:
We will make this change. Default streams will be picked up from the `videos/` folder where user-provided videos may be provided.

Reply:
Will remove.